src/wovnio/wovnphp/custom_domain/CustomDomainLangs.php

```diff
@@ -43,9 +66,52 @@ public function getCustomDomainLangByUrl($url)
         return count($results) <= 0 ? null : array_shift($results);
     }
 
+    public function toHtmlSwapperHash()
+    {
+        $result = array();
+        foreach ($this->customDomainLangs as $lang) {
+            $result[$lang->getHostAndPathWithoutTrailingSlash()] = $lang->getLang();
+        }
+        return $result;
+    }
+
+    /**
+     * Returns the computed (virtual) uri representation for a given physical location uri.
+     * Used when communicating with html-swapper.
+     *
+     * @param $physicalUri string the current uri - pointing to physical location of current lang
+     * @param $lang string lang code of the current uri
+     * @param $defaultLang string lang code of the default (source) language
+     * @return string|string[]|null
+     */
+    public function computeSourceVirtualUrl($physicalUri, $lang, $defaultLang)
+    {
+        $currentLangDomainLang = $this->getSourceCustomDomainByLang($lang);
+        $defaultCustomDomainLang = $this->getCustomDomainLangByLang($defaultLang);
+        return CustomDomainLangUrlHandler::changeToNewCustomDomainLang($physicalUri, $currentLangDomainLang, $defaultCustomDomainLang);
+    }
+
     // parse_url needs protocol to parse URL.
     private function addProtocolIfNeeded($url)
     {
         return preg_match("/https?:\/\//", $url, $matches) ? $url : 'http://' . $url;
     }
+
+    private static function getDefaultLangCustomDomain($customDomainLangsSettingsArray, $defaultLang)
+    {
+        foreach ($customDomainLangsSettingsArray as $langUrl => $config) {
+            if (is_array($config)) {
+                $source = array_key_exists('source', $config) ? $config['source'] : $langUrl;
+                $lang = $config['lang'];
+            } else {
+                $lang = $config;
+                $source = $langUrl;
+            }
+
+            if ($lang === $defaultLang) {
```

Comment:
This feels a bit backwards to me. What do you think about keying the hash by language instead?
```
        $this->customDomainLangsSetting = array(
            'foo.com/' => 'fr',
            'foo.com/path' => array('lang' => 'ja', 'source' => 'japan.foo.com/'),
            'foo.com/dir/path' => 'zh-CHS',
            'english.foo.com/' => array('lang' => 'en', 'source' => 'global.foo.com/')
        );

        $this->customDomainLangsSetting = array(
			'fr' => 'foo.com',
			'ja' => array('url' => 'foo.com/path', 'source' => 'japan.foo.com'),
			'zh-CHS' => 'foo.com/dir/path',
			'en' => array('url' => 'english.foo.com', 'source' => 'global.foo.com')
        );
```

Reply:
I feel we do more lookup based on lang code so I guess that's good.

Reply:
note that I think the json format should also follow this if we do it.

Reply:
If we do that there's the drawback that our json format and ini format no longer matches each other.

But if we decide to do that, can I suggest to group other info together under the language key too? Like I proposed earlier:

```
{
    "langauges": {
        "en": {
            "is_source": false,
            "physical_url": null,
            "url": {
                "domain": "global.example.com",
                "path": null,
                "query": {
                    "key": "lang",
                    "value": "en"
                }
            }
        },
        "fr": {
            "is_source": false,
            "physical_url": null,
            "url": {
                "domain": "global.example.com",
                "path": null,
                "query": {
                    "key": "lang",
                    "value": "fr"
                }
            }
        },
        "ja": {
            "is_source": true,
            "physical_url": {
                "domain": "example.co.jp",
                "path": null,
                "query": null
            },
            "url": {
                "domain": "example.ne.jp",
                "path": null,
                "query": null
            }
        },
        "cn": {
            "is_source": true,
            "physical_url": null,
            "url": {
                "domain": "example.ne.jp",
                "path": "/cn/",
                "query": null
            }
        }
    }
    
}
```

Reply:
Custom domain isn't supported at all in the ini format I thought (or at least we shouldn't support it)

Reply:
We do support "custom domain classic" in the old ini format, there are special handling in the code to make that happen.

Reply:
I think that kind of format might be going too far right now. I would save it for if we decide to overhaul the url pattern system.

Reply:
Yeah, I would remove support for custom domain classic in this change 😉

Reply:
I mean why don't we just keep quiet about this feature and just push JSON to anyone who wants custom domain, because... we don't exactly have a good way to make PHP stop service? Like, we don't really have a way to "remove" support unless we check specifically when reading from .ini for that keyword and raise a fatal error.

Reply:
Sorry, I don't really mean remove support in the code I guess. It is now undefined behavior to add a custom_domain_langs property to your ini file. Done? Everyone who is going to use this feature will be instructed to use the json format 😄

Reply:
I'll add that into the documentation for WOVN.php.